octree/include/pcl/octree/octree_iterator.h

```diff
@@ -132,10 +132,10 @@ namespace pcl
         {
           return (this == &other) ||
                     ((octree_ == other.octree_) &&
-                     (max_octree_depth_ == other.max_octree_depth_) &&
-                     ((current_state_ == other.current_state_) || // end state case
-                      (current_state_ && other.current_state_ &&  // null dereference protection
-                        (current_state_->key_ == other.current_state_->key_))));
+                     ((!(current_state_) && !(other.current_state_)) || // end state case
+                      ((max_octree_depth_ == other.max_octree_depth_) &&
+                       (current_state_ && other.current_state_ &&  // null dereference protection
+                        (current_state_->key_ == other.current_state_->key_)))));
```

Comment:
Am I the only one who finds this return statement extremely convoluted and confusing? What about something like this? I suppose the compiler will produce the same assembly anyway.

```cpp
if (this == &other)  // same object
  return true;
if (octree_ != other.octree_)  // refer to different octrees
  return false;
if (!current_state_ && !other.current_state_)  // both are end iterators
  return true;
if (max_octree_depth_ == other.max_octree_depth_ &&
    current_state_ && other.current_state_ &&  // null dereference protection
    current_state_->key_ == other.current_state_.key_)
  return true;
return false;
```